File: frontend/src/data/seed.ts
import type {
  Job,
  Interview,
  OrgUser,
  CandidateUser,
  Tenant,
} from "./types";

// Must match SEED_TENANT_ID in the backend's app/deps.py — the one seed
// tenant every M6-Phase-1 migration/seed script agrees on.
export const currentTenant: Tenant = {
  id: "11111111-1111-1111-1111-111111111111",
  slug: "northwind-health",
  name: "Northwind Health",
};

export const orgUser: OrgUser = {
  name: "Riley Hoffman",
  email: "riley@northwindhealth.com",
  company: "Northwind Health",
  role: "recruiter",
};

export const candidateUser: CandidateUser = {
  name: "Sophia Martinez",
  email: "sophia.martinez@gmail.com",
};

const backendRubric: Job["rubric"] = [
  {
    id: "c1",
    label: "5+ years backend engineering experience",
    description: "Track record building and operating production backend systems at scale.",
    tag: "Must-have",
    category: "Experience",
    weight: 20,
  },
  {
    id: "c2",
    label: "Proficiency in Go or a comparable statically-typed language",
    description: "Must demonstrate production experience, not just familiarity.",
    tag: "Must-have",
    category: "Technical Skills",
    weight: 20,
  },
  {
    id: "c3",
    label: "Kubernetes & container orchestration",
    description: "Hands-on operating Kubernetes clusters in production.",
    tag: "Must-have",
    category: "Technical Skills",
    weight: 15,
  },
  {
    id: "c4",
    label: "Distributed systems design experience",
    description: "Has designed or scaled multi-service distributed systems.",
    tag: "Nice-to-have",
    category: "Technical Skills",
    weight: 15,
  },
  {
    id: "c5",
    label: "Experience mentoring or leading engineers",
    description: "Tech-lead or mentorship experience preferred, not required.",
    tag: "Nice-to-have",
    category: "Soft Skills",
    weight: 10,
  },
  {
    id: "c6",
    label: "Kafka or equivalent streaming platform",
    description: "Bonus if familiar with event-driven architectures.",
    tag: "Nice-to-have",
    category: "Technical Skills",
    weight: 10,
  },
  {
    id: "c7",
    label: "No unexplained employment gaps > 1 year",
    description: "Flag for recruiter review if present; not an automatic reject.",
    tag: "Disqualifying",
    category: "Other",
    weight: 10,
  },
];

export const jobs: Job[] = [
  {
    id: "job-backend",
    title: "Senior Backend Engineer",
    department: "Engineering",
    location: "Remote",
    type: "Full-time",
    status: "Open",
    description:
      "We're looking for a Senior Backend Engineer to help scale our core platform services.",
    rubric: backendRubric,
    versions: [
      { version: 2, label: "Version 2", status: "Approved", by: "Riley Hoffman", date: "Aug 3, 2026" },
      { version: 1, label: "Version 1", status: "Superseded", by: "Initial draft from JD", date: "Jul 28, 2026" },
    ],
    createdAt: "2026-07-28",
  },
  {
    id: "job-designer",
    title: "Product Designer",
    department: "Design",
    location: "San Francisco, CA",
    type: "Full-time",
    status: "Open",
    description:
      "Own end-to-end product design for our hiring platform, from research to high-fidelity UI.",
    rubric: [
      { id: "d1", label: "3+ years product design experience", description: "B2B SaaS preferred.", tag: "Must-have", category: "Experience", weight: 25 },
      { id: "d2", label: "Strong portfolio of shipped work", description: "Case studies with measurable outcomes.", tag: "Must-have", category: "Portfolio", weight: 30 },
      { id: "d3", label: "Figma fluency", description: "Comfortable building and maintaining design systems.", tag: "Must-have", category: "Technical Skills", weight: 20 },
      { id: "d4", label: "Front-end collaboration experience", description: "Has worked closely with engineers on implementation.", tag: "Nice-to-have", category: "Technical Skills", weight: 15 },
      { id: "d5", label: "User research experience", description: "Has run interviews or usability studies.", tag: "Nice-to-have", category: "Research", weight: 10 },
    ],
    versions: [
      { version: 1, label: "Version 1", status: "Approved", by: "Riley Hoffman", date: "Jul 15, 2026" },
    ],
    createdAt: "2026-07-10",
  },
  {
    id: "job-sre",
    title: "Staff SRE",
    department: "Infrastructure",
    location: "Remote",
    type: "Full-time",
    status: "Closed",
    description: "Lead reliability engineering across our production infrastructure.",
    rubric: [
      { id: "s1", label: "7+ years SRE / infrastructure experience", description: "Staff-level scope of ownership.", tag: "Must-have", category: "Experience", weight: 25 },
      { id: "s2", label: "On-call incident leadership", description: "Has led incident response at scale.", tag: "Must-have", category: "Technical Skills", weight: 20 },
      { id: "s3", label: "Terraform / IaC expertise", description: "Deep infra-as-code experience.", tag: "Must-have", category: "Technical Skills", weight: 20 },
      { id: "s4", label: "Cross-team technical leadership", description: "Has influenced infra strategy org-wide.", tag: "Nice-to-have", category: "Soft Skills", weight: 20 },
      { id: "s5", label: "Cost optimization experience", description: "Track record reducing cloud spend.", tag: "Nice-to-have", category: "Technical Skills", weight: 15 },
    ],
    versions: [
      { version: 1, label: "Version 1", status: "Approved", by: "Riley Hoffman", date: "Jun 1, 2026" },
    ],
    createdAt: "2026-05-20",
  },
  {
    id: "job-marketing",
    title: "Marketing Manager",
    department: "Marketing",
    location: "New York, NY",
    type: "Full-time",
    status: "Draft",
    description: "Drive demand generation and brand strategy for our hiring platform.",
    rubric: [
      { id: "m1", label: "5+ years B2B SaaS marketing", description: "Demand gen and lifecycle experience.", tag: "Must-have", category: "Experience", weight: 30 },
      { id: "m2", label: "Content & positioning ownership", description: "Has owned messaging and positioning.", tag: "Must-have", category: "Skills", weight: 25 },
      { id: "m3", label: "Analytics fluency", description: "Comfortable with funnel and attribution data.", tag: "Nice-to-have", category: "Skills", weight: 20 },
    ],
    versions: [],
    createdAt: "2026-08-06",
  },
];


export const interviews: Interview[] = [
  {
    id: "int-backend-tech",
    title: "Senior Backend Engineer — Technical Screen",
    jobTitle: "Senior Backend Engineer",
    jobId: null,
    candidateId: null,
    candidateName: null,
    mode: "Chat",
    status: "Active",
    duration: 45,
    createdAt: "2026-07-29",
    shared: true,
    questions: [
      { id: "q1", prompt: "Walk me through how you'd design a rate limiter for a public API.", type: "System Design", difficulty: "Medium" },
      { id: "q2", prompt: "Describe a production incident you led the response for. What was the root cause?", type: "Behavioral", difficulty: "Medium" },
      { id: "q3", prompt: "How would you shard a PostgreSQL database that's outgrown a single instance?", type: "Technical", difficulty: "Hard" },
      { id: "q4", prompt: "Tell me about a time you disagreed with a technical decision on your team.", type: "Culture", difficulty: "Easy" },
    ],
  },
  {
    id: "int-designer-portfolio",
    title: "Product Designer — Portfolio Walkthrough",
    jobTitle: "Product Designer",
    jobId: null,
    candidateId: null,
    candidateName: null,
    mode: "Voice",
    status: "Active",
    duration: 30,
    createdAt: "2026-07-20",
    shared: false,
    questions: [
      { id: "q1", prompt: "Walk me through a project where the initial design didn't work and how you iterated.", type: "Behavioral", difficulty: "Medium" },
      { id: "q2", prompt: "How do you decide when a design is ready to ship vs. needs more research?", type: "Culture", difficulty: "Easy" },
    ],
  },
  {
    id: "int-sre-oncall",
    title: "Staff SRE — Incident Leadership",
    jobTitle: "Staff SRE",
    jobId: null,
    candidateId: null,
    candidateName: null,
    mode: "Avatar",
    status: "Archived",
    duration: 40,
    createdAt: "2026-06-05",
    shared: true,
    questions: [
      { id: "q1", prompt: "Describe the worst production outage you've handled. Walk me through your decisions minute by minute.", type: "Behavioral", difficulty: "Hard" },
      { id: "q2", prompt: "How would you reduce our cloud spend by 20% without impacting reliability?", type: "Technical", difficulty: "Medium" },
    ],
  },
];

File: app/seed.py
"""Seed the database from the generated frontend seed.

Idempotent: skips jobs/candidates/interviews that already exist (matched by
original synthetic id via an "original_id" side-table? no — matched by title/email).

Writes the canonical seed (37 jobs, 228 candidate applications, 2 interviews)
into Postgres so the API serves the same data the frontend used to hold in
localStorage.

Usage: venv/bin/python -m app.seed
"""

from __future__ import annotations

import asyncio
import json
import re
from pathlib import Path

from sqlalchemy import select

from app.db import async_session
from app.models.application import Application
from app.models.candidate import Candidate
from app.models.interview import Interview
from app.models.job import Job
from app.models.user import User

SEED_TS = Path(__file__).resolve().parents[1] / "frontend" / "src" / "data" / "generated-seed.ts"

SEED_INTERVIEWS = [
    {
        "title": "Senior Backend Engineer — Technical Screen",
        "jobTitle": "Senior Backend Engineer",
        "mode": "Chat",
        "status": "Active",
        "duration": 45,
        "shared": True,
        "questions": [
            {"id": "q1", "prompt": "Walk me through how you'd design a rate limiter for a public API.", "type": "System Design", "difficulty": "Medium"},
            {"id": "q2", "prompt": "Describe a production incident you led the response for. What was the root cause?", "type": "Behavioral", "difficulty": "Medium"},
            {"id": "q3", "prompt": "How would you shard a PostgreSQL database that's outgrown a single instance?", "type": "Technical", "difficulty": "Hard"},
            {"id": "q4", "prompt": "Tell me about a time you disagreed with a technical decision on your team.", "type": "Culture", "difficulty": "Easy"},
        ],
    },
    {
        "title": "Product Designer — Portfolio Walkthrough",
        "jobTitle": "Product Designer",
        "mode": "Voice",
        "status": "Active",
        "duration": 30,
        "shared": False,
        "questions": [
            {"id": "q1", "prompt": "Walk me through a project where the initial design didn't work and how you iterated.", "type": "Behavioral", "difficulty": "Medium"},
            {"id": "q2", "prompt": "How do you decide when a design is ready to ship vs. needs more research?", "type": "Culture", "difficulty": "Easy"},
        ],
    },
    {
        "title": "Staff SRE — Incident Leadership",
        "jobTitle": "Staff SRE",
        "mode": "Avatar",
        "status": "Archived",
        "duration": 40,
        "shared": True,
        "questions": [
            {"id": "q1", "prompt": "Walk me through the largest outage you've led the response for.", "type": "System Design", "difficulty": "Hard"},
            {"id": "q2", "prompt": "How do you balance on-call load across a team of 6?", "type": "Behavioral", "difficulty": "Medium"},
        ],
    },
]


def _parse_ts_array(src: str, name: str) -> list[dict]:
    m = re.search(rf"export const {name}:\s*\w+\[\]\s*=\s*(\[.*?\])\s*;", src, re.S)
    if not m:
        raise RuntimeError(f"could not find export const {name} in {src[:200]}")
    body = m.group(1)
    # The TS literal is valid JSON apart from trailing commas inside objects,
    # which json.loads rejects. Strip trailing commas before object/array close.
    cleaned = re.sub(r",\s*([}\]])", r"\1", body)
    return json.loads(cleaned)


def _load_ts(path: Path) -> tuple[list[dict], list[dict]]:
    src = path.read_text()
    jobs = _parse_ts_array(src, "jobs") if "jobs" in src else []
    candidates = _parse_ts_array(src, "candidates") if "candidates" in src else []
    return jobs, candidates


async def seed() -> None:
    jobs, candidates = _load_ts(SEED_TS)
    interviews = SEED_INTERVIEWS

    async with async_session() as db:
        org = await db.execute(select(User).where(User.email == "riley@northwindhealth.com"))
        org_user = org.scalar_one_or_none()

        seen_emails: dict[str, Candidate] = {}

        for j in jobs:
            existing = (await db.execute(select(Job).where(Job.title == j["title"]))).scalars().first()
            if existing:
                continue
            db.add(
                Job(
                    title=j["title"],
                    department=j.get("department"),
                    location=j.get("location"),
                    employment_type=j.get("type"),
                    status=j.get("status", "Open"),
                    description=j.get("description", ""),
                    rubric=j.get("rubric", []),
                    versions=j.get("versions", []),
                    posted_by=org_user.id if org_user else None,
                )
            )
        await db.commit()
        print(f"Jobs seeded: {len(jobs)}")

        # Resolve the synthetic jobId (a slug) -> Job via the seed title index.
        title_by_id = {j["id"]: j["title"] for j in jobs}
        created = 0
        for c in candidates:
            title = title_by_id.get(c["jobId"])
            if title is None:
                continue
            job = (await db.execute(select(Job).where(Job.title == title))).scalars().first()
            if job is None:
                continue
            person = seen_emails.get(c["email"])
            if person is None:
                person = Candidate(
                    name=c["name"],
                    email=c["email"],
                    phone=c.get("phone"),
                    location=c.get("location"),
                    source=c.get("source", "Manual Entry"),
                    tags=c.get("tags", []),
                    notes=c.get("notes", ""),
                    resume_file=c.get("resumeFile"),
                    years_exp=c.get("yearsExp", 0),
                    current_title=c.get("currentTitle", "—"),
                    current_company=c.get("currentCompany", "—"),
                    skills=c.get("skills", []),
                    summary=c.get("summary", ""),
                    experience=c.get("experience", []),
                    education=c.get("education", "—"),
                    certifications=c.get("certifications", "—"),
                )
                db.add(person)
                await db.flush()
                seen_emails[c["email"]] = person

            dup = (
                await db.execute(
                    select(Application).where(
                        Application.candidate_id == person.id,
                        Application.job_id == job.id,
                    )
                )
            ).scalars().first()
            if dup:
                continue
            db.add(
                Application(
                    candidate_id=person.id,
                    job_id=job.id,
                    status="screening",
                    match_score=c.get("score", 0),
                    shortlisted=c.get("shortlisted", False),
                    decision=c.get("decision", "None"),
                    pipeline_stage=c.get("pipelineStage", "Applied"),
                    scorecard=c.get("scorecard", []),
                    strengths=c.get("strengths", []),
                    gaps=c.get("gaps", []),
                    compare_verdict=c.get("compareVerdict", "Pass"),
                    ai_note=c.get("aiNote", ""),
                    applied_at=_iso(c.get("appliedAt")),
                )
            )
            created += 1
        await db.commit()
        print(f"Candidates seeded: {created} applications across {len(seen_emails)} people")

        for iv in interviews:
            existing = (
                await db.execute(select(Interview).where(Interview.title == iv["title"]))
            ).scalars().first()
            if existing:
                continue
            db.add(
                Interview(
                    title=iv["title"],
                    job_title=iv.get("jobTitle", ""),
                    mode=iv.get("mode", "Chat"),
                    status=iv.get("status", "Draft"),
                    questions=iv.get("questions", []),
                    duration=iv.get("duration", 30),
                    shared=iv.get("shared", False),
                )
            )
        await db.commit()
        print(f"Interviews seeded: {len(interviews)}")


def _iso(value: str | None):
    from datetime import datetime

    if not value:
        return None
    try:
        return datetime.fromisoformat(value)
    except ValueError:
        return None


if __name__ == "__main__":
    asyncio.run(seed())
